File: ruby/food-chain/food_chain.rb
class FoodChainSong
  ANIMALS = {
    fly: '',
    spider: "It wriggled and jiggled and tickled inside her.\n",
    bird: "How absurd to swallow a bird!\n",
    cat: "Imagine that, to swallow a cat!\n",
    dog: "What a hog, to swallow a dog!\n",
    goat: "Just opened her throat and swallowed a goat!\n",
    cow: "I don't know how she swallowed a cow!\n",
    horse: "She's dead, of course!\n"
  }.freeze

  def verse(verse_number)
    animal_index = verse_number - 1
    first_animal = ANIMALS.keys[animal_index]
    animal_chain = ANIMALS.keys[0..animal_index].reverse

    lyrics = "I know an old lady who swallowed a #{first_animal}.\n"
    lyrics << ANIMALS[first_animal]

    return lyrics if first_animal == :horse

    animal_chain.each_cons(2) do |a1, a2|
      if a2 == :spider
        a2 = 'spider that wriggled and jiggled and tickled inside her'
      end

      lyrics << "She swallowed the #{a1} to catch the #{a2}.\n"
    end

    lyrics << "I don't know why she swallowed the fly. Perhaps she'll die.\n"
  end

  def verses(from, to)
    (from..to).inject('') do |lyrics, verse|
      lyrics << verse(verse)
      lyrics << "\n"
    end
  end

  def sing
    verses(1, 8)
  end
end

File: go/food-chain/food_chain.go
package foodchain

import (
	"bytes"
)

// TestVersion is the version of the test suite this code is written to.
const TestVersion = 1

var animals = []string{
	"fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse",
}

var animalDescription = []string{
	"",
	"It wriggled and jiggled and tickled inside her.\n",
	"How absurd to swallow a bird!\n",
	"Imagine that, to swallow a cat!\n",
	"What a hog, to swallow a dog!\n",
	"Just opened her throat and swallowed a goat!\n",
	"I don't know how she swallowed a cow!\n",
	"She's dead, of course!",
}

// Verse returns the specified verse of the song 'I Know an Old Lady Who Swallowed a Fly'
func Verse(verse int) string {
	animalIndex := verse - 1
	firstAnimal := animals[animalIndex]
	animalChain := animals[0 : animalIndex+1]
	var lyrics bytes.Buffer

	lyrics.WriteString("I know an old lady who swallowed a ")
	lyrics.WriteString(firstAnimal)
	lyrics.WriteString(".\n")
	lyrics.WriteString(animalDescription[animalIndex])

	if firstAnimal == "horse" {
		return lyrics.String()
	}

	for i := len(animalChain) - 1; i > 0; i-- {
		animal1 := animalChain[i]
		animal2 := animalChain[i-1]

		if animal2 == "spider" {
			animal2 = "spider that wriggled and jiggled and tickled inside her"
		}

		lyrics.WriteString("She swallowed the ")
		lyrics.WriteString(animal1)
		lyrics.WriteString(" to catch the ")
		lyrics.WriteString(animal2)
		lyrics.WriteString(".\n")

	}

	lyrics.WriteString("I don't know why she swallowed the fly. Perhaps she'll die.")

	return lyrics.String()
}

// Verses returns the verses between from and to of the song 'I Know an Old Lady Who Swallowed a Fly'
func Verses(from, to int) string {
	var lyrics bytes.Buffer

	for i := from; i < to; i++ {
		lyrics.WriteString(Verse(i))
		lyrics.WriteString("\n\n")
	}
	lyrics.WriteString(Verse(to))

	return lyrics.String()
}

// Song returns all the verses of the song 'I Know an Old Lady Who Swallowed a Fly'
func Song() string {
	return Verses(1, 8)
}
